File: 0491-non-decreasing-subsequences/0491-non-decreasing-subsequences.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<int> path;
    vector<vector<int>> findSubsequences(vector<int>& nums) {
        backtracking(nums, 0);
        return res;
    }

    void backtracking(vector<int>& nums, int start) {
        if (path.size() > 1) {
            vector<int> copy = path;
            res.push_back(copy);
        }
        
        unordered_set<int> set;
        for (int i = start; i < nums.size(); i++) {
            if (set.find(nums[i]) != set.end()) {
                continue;
            }
            set.insert(nums[i]);
            if (path.empty() || nums[i] >= path[path.size() - 1]) {
                path.push_back(nums[i]);
                backtracking(nums, i + 1);
                path.pop_back();
            }
        }
    }
};

File: 0491-non-decreasing-subsequences/0491-non-decreasing-subsequences.py
class Solution:
    def __init__(self):
        self.subset = []
        self.result = []

    def findSubsequences(self, nums: List[int]) -> List[List[int]]:
        self.used = [False] * len(nums)
        self.backtracking(nums, 0)
        return self.result
        
    def backtracking(self, nums: int, start: int):
        if len(self.subset) > 1:
            self.result.append(copy.copy(self.subset))

        used = set()
        for i in range(start, len(nums)):
            if nums[i] in used:
                continue

            if not self.subset or nums[i] >= self.subset[-1]:
                used.add(nums[i])
                self.subset.append(nums[i])
                self.backtracking(nums, i + 1)
                self.subset.pop()